File: 209-minimum-size-subarray-sum/209-minimum-size-subarray-sum.cpp
class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        return slidingWindow(target, nums);
    }
private:
    // O(n) time and O(1) extra space
    int slidingWindow(int target, vector<int>& nums) {
        int wStart = 0, minSize = INT_MAX, wSum = 0;
        
        for(int wEnd = 0; wEnd < nums.size(); ++wEnd) {
            wSum += nums[wEnd];
            
            while(wSum >= target) {
                wSum -= nums[wStart];
                minSize = min(minSize, wEnd - wStart + 1);
                ++wStart;
            }
        }
        if(minSize >= nums.size()+1)
            return 0;
        return minSize;
    }
};

File: 209-minimum-size-subarray-sum/209-minimum-size-subarray-sum.py
class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        
        # O(n) Time and O(1) extra space
        wStart, wSum, minSize = 0, 0, len(nums) + 1
        
        for wEnd in range(len(nums)):
            wSum += nums[wEnd]
            
            while wSum >= target:
                wSum -= nums[wStart]
                minSize = min(minSize, wEnd - wStart + 1)
                wStart += 1
        if minSize >= len(nums) + 1:
            return 0
        return minSize